crypto/hpke/hpke.c

```diff
@@ -0,0 +1,3137 @@
+ */
+
+/*
+ * @brief do the AEAD decryption
+ *
+ * @param libctx is the context to use (normally NULL)
+ * @param suite is the ciphersuite
+ * @param key is the secret
+ * @param keylen is the length of the secret
+ * @param iv is the initialisation vector
+ * @param ivlen is the length of the iv
+ * @param aad is the additional authenticated data
+ * @param aadlen is the length of the aad
+ * @param cipher is obvious
+ * @param cipherlen is the ciphertext length
+ * @param plain is an output
+ * @param plainlen input/output, better be big enough on input, exact on output
+ * @return 1 for good otherwise bad
+ */
+static int hpke_aead_dec(OSSL_LIB_CTX *libctx,
+                         OSSL_HPKE_SUITE suite,
+                         unsigned char *key, size_t keylen,
+                         unsigned char *iv, size_t ivlen,
+                         unsigned char *aad, size_t aadlen,
+                         unsigned char *cipher, size_t cipherlen,
+                         unsigned char *plain, size_t *plainlen)
+{
+    int erv = 1;
+    EVP_CIPHER_CTX *ctx = NULL;
+    int len = 0;
+    size_t plaintextlen = 0;
+    unsigned char *plaintext = NULL;
+    size_t taglen;
+    uint16_t aead_ind = 0;
+    EVP_CIPHER *enc = NULL;
+
+    aead_ind = aead_iana2index(suite.aead_id);
+    if (aead_ind == 0) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    taglen = hpke_aead_tab[aead_ind].taglen;
+    plaintext = OPENSSL_malloc(cipherlen);
+    if (plaintext == NULL) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    /* Create and initialise the context */
+    if ((ctx = EVP_CIPHER_CTX_new()) == NULL) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    /* Initialise the encryption operation */
+    enc = EVP_CIPHER_fetch(libctx, hpke_aead_tab[aead_ind].name, NULL);
+    if (enc == NULL) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    if (EVP_DecryptInit_ex(ctx, enc, NULL, NULL, NULL) != 1) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    EVP_CIPHER_free(enc);
+    enc = NULL;
+    if (EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN, ivlen, NULL) != 1) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    /* Initialise key and IV */
+    if (EVP_DecryptInit_ex(ctx, NULL, NULL, key, iv) != 1) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    /* Provide AAD. Can be called zero or more times as required */
+    if (aadlen != 0 && aad != NULL) {
+        if (EVP_DecryptUpdate(ctx, NULL, &len, aad, aadlen) != 1) {
+            OSSL_HPKE_err;
+            goto err;
+        }
+    }
+    /*
+     * Provide the message to be decrypted, and obtain cleartext output.
+     * EVP_DecryptUpdate can be called multiple times if necessary
+     */
+    if (EVP_DecryptUpdate(ctx, plaintext, &len, cipher,
+                          cipherlen - taglen) != 1) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    plaintextlen = len;
+    /* Set expected tag value. Works in OpenSSL 1.0.1d and later */
```

Comment:
The "works in" comment isn't really relevant. This code will go into master only, so its irrelevant what old versions this might work for.